### ISSUE TITLE
Comm Engine and 'external' tags registration

### ISSUE THREAD

**Issue by therault** [ICLDisco/parsec#467]
## Describe the bug

registering tags higher than `PARSEC_CE_REMOTE_DEP_MAX_CTRL_TAG` but lower than PARSEC_MAX_REGISTERED_TAGS and lower than `MPI_FUNNELLED_MAX_TAG` results in mis-matching, and eventually in a MPI_ERR_TRUNCATE error. I don't know if it is the expected behavior or not. The comments in parsec_mpi_funnelled.c made me thin it should be authorized, but it didn't work.

### To Reproduce

ttg, the new branch on top of parsec master, can be used to reproduce: use tags 10 and 11 to implement `parsec_ttg_tag` and `parsec_ttg_rma_tag` respectively 

### Expected behavior

I would have expected the system to accept tags above `PARSEC_CE_REMOTE_DEP_MAX_CTRL_TAG`, I thought `PARSEC_CE_REMOTE_DEP_MAX_CTRL_TAG` was only for the pre-defined tags. If that's the expected behavior, I'm fine with it, but we should  document that more clearly somewhere.

## Environment (please complete the following information):

I will put the link to the ttg branch when it's up.

## Additional context
The content of the `config.log` file can be useful in some cases.

**Comment by therault** [comment]
The TTG branch to test is https://github.com/therault/ttg/tree/ttg-over-parsec-master

**Comment by bosilca** [comment]
I think the root cause of this truncation error is due to a bug in the generation of the data transfer tags. We are using `PARSEC_CE_REMOTE_DEP_MAX_CTRL_TAG` instead of `PARSEC_MAX_REGISTERED_TAGS`, and this enable collisions in the range `[PARSEC_CE_REMOTE_DEP_MAX_CTRL_TAG+1 .. PARSEC_MAX_REGISTERED_TAGS].

Please try with the following patch. If it fixes the issue, let me know and I will prepare a quick PR.

```diff
diff --git a/parsec/parsec_mpi_funnelled.c b/parsec/parsec_mpi_funnelled.c
index 3fb4e3ee6..b6a93df2a 100644
--- a/parsec/parsec_mpi_funnelled.c
+++ b/parsec/parsec_mpi_funnelled.c
@@ -57,13 +57,15 @@ PARSEC_OBJ_CLASS_INSTANCE(mpi_funnelled_mem_reg_handle_t, parsec_list_item_t,
 #error "unsigned long is not large enough to hold a pointer!"
 #endif

-/* note: tags are necessary to order communication between pairs. They are used to
- * correctly handle data transfers, as each data provider will provide a tag which
+/* note: tags are necessary to match data communication between pairs. They are used to
+ * correctly handle data transfers, as each data provider will also provide a tag which
  * combined with the source ensure message matching consistency. As MPI requires the
  * max tag to be positive, initializing it to a negative value allows us to check
  * if the layer has been initialized or not.
+ *
+ * All data movements tags should be outside the range allowed for AM tags.
  */
-#define MIN_MPI_TAG (PARSEC_CE_REMOTE_DEP_MAX_CTRL_TAG+1)
+#define MIN_MPI_TAG (PARSEC_MAX_REGISTERED_TAGS+1)
 static int MAX_MPI_TAG = -1, mca_tag_ub = -1;
 static volatile int __VAL_NEXT_TAG = MIN_MPI_TAG;
 #if INT_MAX == INT32_MAX
```

**Comment by devreal** [comment]
IMO, the right way to fix this is to have the comm engine manage the tag space. That is the only way to avoid conflicts in the long run.

So instead of

```
parsec_ce.tag_register(PARSEC_TERMDET_FOURCOUNTER_MSG_TAG, parsec_termdet_fourcounter_msg_dispatch, ptr,
                                          PARSEC_TERMDET_FOURCOUNTER_MAX_MSG_SIZE);
```

with `PARSEC_TERMDET_FOURCOUNTER_MSG_TAG` being defined by the comm engine (!) we should do:


```
int tag;
parsec_ce.tag_register(&tag, parsec_termdet_fourcounter_msg_dispatch, ptr,
                                          PARSEC_TERMDET_FOURCOUNTER_MAX_MSG_SIZE);
```

The approach of hoping that all users of the comm engine play nice and agree on sub-spaces that don't overlap is too fragile. Tag registration is part of the setup so it's serial, i.e., all processes will have the same value for the same semantic tag. I'm happy to provide a patch for it if we can agree on that change :)

**Comment by omor1** [comment]
> IMO, the right way to fix this is to have the comm engine manage the tag space. That is the only way to avoid conflicts in the long run.

I concur with this.

**Comment by bosilca** [comment]
That's a totally different discussion. The correct way to fix this issue is to avoid the automatic tag generation from colliding with the predefined tags. We can talk about dynamic tags at a later date assuming we have legitimate scenario where we need them.

**Comment by bosilca** [comment]
I don't see how having dynamic tags will avoid tag collision between AM and data transfers. The issue here is that the tag computation was using the wrong define, allowing tags into a range that should have been dedicated to AM. Anyway, I kind of like the multiple communicator approach, it gives a much cleaner implementation. I will prepare a PR, meanwhile this quick fix should be address the issue.

**Comment by therault** [comment]
Sorry to be late in the discussion. The patch looks fine: I tried by using two tags claimed by TTG (10 and 11 as before, instead of 8 and 9), and those tags are not part of the pre-defined tag set known in parsec, and the code works as expected in sequential and in distributed.

A simpler solution than PR #468 that uses a function pointer registration function, in which the callback functions call other function pointers would be still welcome (independently of this fix). :)

> I think the root cause of this truncation error is due to a bug in the generation of the data transfer tags. We are using `PARSEC_CE_REMOTE_DEP_MAX_CTRL_TAG` instead of `PARSEC_MAX_REGISTERED_TAGS`, and this enable collisions in the range `[PARSEC_CE_REMOTE_DEP_MAX_CTRL_TAG+1 .. PARSEC_MAX_REGISTERED_TAGS]`.
> 
> Please try with the following patch. If it fixes the issue, let me know and I will prepare a quick PR.
> 
> ```diff
> diff --git a/parsec/parsec_mpi_funnelled.c b/parsec/parsec_mpi_funnelled.c
> index 3fb4e3ee6..b6a93df2a 100644
> --- a/parsec/parsec_mpi_funnelled.c
> +++ b/parsec/parsec_mpi_funnelled.c
> @@ -57,13 +57,15 @@ PARSEC_OBJ_CLASS_INSTANCE(mpi_funnelled_mem_reg_handle_t, parsec_list_item_t,
>  #error "unsigned long is not large enough to hold a pointer!"
>  #endif
> 
> -/* note: tags are necessary to order communication between pairs. They are used to
> - * correctly handle data transfers, as each data provider will provide a tag which
> +/* note: tags are necessary to match data communication between pairs. They are used to
> + * correctly handle data transfers, as each data provider will also provide a tag which
>   * combined with the source ensure message matching consistency. As MPI requires the
>   * max tag to be positive, initializing it to a negative value allows us to check
>   * if the layer has been initialized or not.
> + *
> + * All data movements tags should be outside the range allowed for AM tags.
>   */
> -#define MIN_MPI_TAG (PARSEC_CE_REMOTE_DEP_MAX_CTRL_TAG+1)
> +#define MIN_MPI_TAG (PARSEC_MAX_REGISTERED_TAGS+1)
>  static int MAX_MPI_TAG = -1, mca_tag_ub = -1;
>  static volatile int __VAL_NEXT_TAG = MIN_MPI_TAG;
>  #if INT_MAX == INT32_MAX
> ```

**Comment by bosilca** [comment]
This has been fixed by #477 (separating the AM and data transfers into distinct communicators).
